bson/value_reader_test.go

```diff
@@ -1347,13 +1484,303 @@ func TestValueReader(t *testing.T) {
 					vr := &valueReader{
-						r: bufio.NewReader(bytes.NewReader(tc.data[tc.startingOffset:tc.offset])),
+						src: &bufferedValueReader{buf: tc.data, offset: tc.startingOffset},
 						stack: []vrState{
 							{mode: mTopLevel, end: startingEnd},
 							{mode: mElement, vType: tc.t},
 						},
-						frame:  1,
-						offset: tc.startingOffset,
+						frame: 1,
+					}
+
+					err := vr.Skip()
+					if !errequal(t, err, tc.err) {
+						t.Errorf("Did not receive expected error; got %v; want %v", err, tc.err)
+					}
+					if tc.err == nil && vr.src.pos() != tc.offset {
+						t.Errorf("Offset not set at correct position; got %d; want %d", vr.src.pos(), tc.offset)
+					}
+				})
+				t.Run("ReadBytes", func(t *testing.T) {
+					vr := &valueReader{
+						src: &bufferedValueReader{buf: tc.data, offset: tc.startingOffset},
+						stack: []vrState{
+							{mode: mTopLevel, end: startingEnd},
+							{mode: mElement, vType: tc.t},
+						},
+						frame: 1,
+					}
+
+					_, got, err := vr.readValueBytes(nil)
+					if !errequal(t, err, tc.err) {
+						t.Errorf("Did not receive expected error; got %v; want %v", err, tc.err)
+					}
+					if tc.err == nil && vr.src.pos() != tc.offset {
+						t.Errorf("Offset not set at correct position; got %d; want %d", vr.src.pos(), tc.offset)
+					}
+					if tc.err == nil && !bytes.Equal(got, tc.data[tc.startingOffset:]) {
+						t.Errorf("Did not receive expected bytes. got %v; want %v", got, tc.data[tc.startingOffset:])
+					}
+				})
+			})
+		}
+		t.Run("ReadValueBytes/Top Level Doc", func(t *testing.T) {
+			testCases := []struct {
+				name     string
+				want     []byte
+				wantType Type
+				wantErr  error
+			}{
+				{
+					"success",
+					bsoncore.BuildDocument(nil, bsoncore.AppendDoubleElement(nil, "pi", 3.14159)),
+					Type(0),
+					nil,
+				},
+				{
+					"wrong length",
+					[]byte{0x01, 0x02, 0x03},
+					Type(0),
+					io.EOF,
+				},
+				{
+					"append bytes",
+					[]byte{0x01, 0x02, 0x03, 0x04},
+					Type(0),
+					io.EOF,
+				},
+			}
+
+			for _, tc := range testCases {
+				tc := tc
+				t.Run(tc.name, func(t *testing.T) {
+					t.Parallel()
+					vr := &valueReader{
+						src: &bufferedValueReader{buf: tc.want},
+						stack: []vrState{
+							{mode: mTopLevel},
+						},
+						frame: 0,
+					}
+					gotType, got, gotErr := vr.readValueBytes(nil)
+					if !errors.Is(gotErr, tc.wantErr) {
+						t.Errorf("Did not receive expected error. got %v; want %v", gotErr, tc.wantErr)
+					}
+					if tc.wantErr == nil && gotType != tc.wantType {
+						t.Errorf("Did not receive expected type. got %v; want %v", gotType, tc.wantType)
+					}
+					if tc.wantErr == nil && !bytes.Equal(got, tc.want) {
+						t.Errorf("Did not receive expected bytes. got %v; want %v", got, tc.want)
+					}
+				})
+			}
+		})
+	})
+
+	// This test is too complicated to try to abstract using subtests.
+	t.Run("ReadBytes & Skip (streaming)", func(t *testing.T) {
```

Comment:
Optional: Consider making this a separate test function to make the test code easier to read and run.